### PR TITLE
Issue 10306: Improve local delivery

### LINKED ISSUE #10306: Followers not seeing my reshares: is this expected behavior?
Specifically, they can see them when viewing my profile directly, but they don't show up in their network feeds.

Didn't want to start debugging before confirming this behavior is not intended.

### PR DESCRIPTION
**Pull request by annando**
This fixes #10306

We had several issues with the delivery of activities on the same machine. This is now completely reworked so that the behaviour should be idential to a communication between two Friendica servers (using AP).

Some tests are still open, for example with local forums. But I guess this can be done until the release is done.

### CODE REVIEW

**Comment by AlfredSK** [comment]
Do you mean sending posts to a local forum and looking if other local users get the forum posts (if they are a follower of the forum)?
I can test this once this has been merged.

**Comment by annando** [comment]
Yeah, stuff like this.

File: src/Model/Item.php [hunk 1]
```diff
@@ -1322,19 +1346,26 @@ public static function distribute($itemid, $signed_text = '')
...
+	 * @param array   $fields     Additional fields to be stored
+	 * @param integer $source_uid User id of the source post
 	 * @return integer stored item id
 	 */
-	public static function storeForUserByUriId(int $uri_id, int $uid, array $fields = [])
+	public static function storeForUserByUriId(int $uri_id, int $uid, array $fields = [], int $source_uid = 0)
 	{
-		$item = Post::selectFirst(self::ITEM_FIELDLIST, ['uri-id' => $uri_id, 'uid' => 0]);
+		if ($uid == $source_uid) {
+			Logger::warning('target UID must be be equal to the source UID', ['uri-id' => $uri_id, 'uid' => $uid]);
```
Comment (MrPetovan) [inline src/Model/Item.php]: The log message contradicts the condition.

File: src/Model/Item.php [hunk 2]
```diff
@@ -1364,11 +1411,21 @@ private static function storeForUser(array $item, int $uid)
...
 		unset($item['starred']);
 		unset($item['unseen']);
 		unset($item['psid']);
+		unset($item['pinned']);
+		unset($item['ignored']);
+		unset($item['pubmail']);
+		unset($item['forum_mode']);
+
+		//unset($item['post-reason']);
+		//unset($item['protocol']);
```
Comment (MrPetovan) [inline src/Model/Item.php]: Please don't add commented out code, instead stash it in git if you need it for later.

File: src/Model/User.php [hunk 3]
```diff
@@ -313,7 +313,7 @@ public static function getByNickname($nickname, array $fields = [])
 	public static function getIdForURL(string $url)
 	{
 		// Avoid any database requests when the hostname isn't even part of the url.
```
Comment (MrPetovan) [inline src/Model/User.php]: Please update this comment to mention the local node hostname.
```suggestion
		// Avoid database queries when the local node hostname isn't even part of the url.
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Model/Item.php (hunk 2 above)
Comment (annando) [reply src/Model/Item.php]: Sorry, I forgot to push to additional commits I did today.

File: src/Model/Item.php (hunk 1 above)
Comment (annando) [reply src/Model/Item.php]: Don't you see the invisible `not` there? 😁

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**